### DISCUSSION TITLE
Deal with dual dynamic imports (Machine + UI Component)

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2344] [Q&A] [answered]
Greetings, thank you a million for this fantastic library that I think will be the future of state management.

I am planning to encapsulate all of business logic and data flow within a single machine with many actors.
Each actor is usually consumed by a corresponding UI component but I keep them in two separate files.
Now, my actors and their UI components are dynamically imported when needed to keep the initial bundle to minimum.
For this reason, I ended up having two dynamic import requests:
1. The first one is issued form within the machine to load the actor, spawn it, then change the state to indicate readiness for use.
2. The second one is issued from within the UI to load the UI component, render it, and feed it the actor instance.

I would like to either eliminate the second request or batch it with the first one, without having to merge the two files or create a third artificial one as barrel?

Many thanks in advance.

**Comment by ChrisShank** [comment, marked as answer]
I think you are sort of at the whim of dynamic imports here! If you are using webpack than you can group chucks together by specifying them with the same `webpackChunkName`. https://webpack.js.org/api/module-methods/#magic-comments

This will only work with webpack though so if you are using another bundler i think creating a third file and dynamically importing that is your best bet!

Also out of curiosity how are you dynamically importing actors!?

**Reply by geekox86** [reply]
`import(...)` returns a promise which can be invoked from my state machine and on success I spawn the dynamically loaded actor then transition to another state that makes use of it.

**Reply by ChrisShank** [reply]
Thats an interesting approach!! I am in the early stages of building out a helper to dynamically import a machine. The tentative implementation can be seen [here](https://discord.com/channels/795785288994652170/799416911259369472/853051596481363970). Happy to hear your thoughts! :)

**Reply by geekox86** [reply]
I could not access the link on Discord.

**Reply by ChrisShank** [reply]
gotcha here is the **tentative** implementation! I think there are some more things that we need to add, but the gist is using invoked callbacks.

```ts
function invokeAsyncMachine(loadMachine: () => Promise<Machine>) {
  return () => (sendBack, receive) => {
    let service = null;
    let status: 'pending' | 'resolved' | 'rejected' | 'stopped' = 'pending';
    const pendingMessages: Event[] = []; // keep track of events sent while the machine is loading

    loadMachine()
      .then(machine => {
        if (status === 'stopped') return; // Return if the invoked state is exited before the machine is loaded 

        status = 'resolved';
        service = interpret(machine, {
          parent: { send: sendBack },
        });
        service.send(pendingMessages);
      })
      .catch(() => status = 'rejected');

    receive(event => {
      if (status === 'pending') {
        pendingMessages.push(event)
      } else if (status === 'resolved' && service) {
        sevice.send(event);
      }
    });

    return () => {
      status = 'stopped';
      service?.stop();
    };
  };
}
```

```ts
const parentMachine = createMachine({
  id: 'parent',
  initial: 'idle',
  states: {
    idle: {
      on: { START: 'active' },
    },
    active: {
      invoke: { 
        id: 'asyncMachine', 
        src: invokeAsyncMachine(() => import('./async-machine')),
      },
      on: { DONE: 'finished' },
    },
    finished: {},
  },
});
```

**Reply by geekox86** [reply]
Well done. I really liked the idea of pending messages. Thx for sharing.
